pkg/migrations/sql/idempotent/exemplar.sql

```diff
@@ -0,0 +1,123 @@
+        ) AS row GROUP BY row.metric_name
+    );
+END;
+$$
+LANGUAGE PLPGSQL;
+GRANT EXECUTE ON FUNCTION SCHEMA_CATALOG.get_exemplar_label_key_positions(TEXT, TEXT[]) TO prom_writer;
+GRANT EXECUTE ON FUNCTION SCHEMA_CATALOG.get_exemplar_label_key_positions(TEXT, TEXT[]) TO prom_reader;
+
+CREATE OR REPLACE FUNCTION SCHEMA_CATALOG.get_exemplar_label_key_positions(metric_name_text TEXT)
+RETURNS JSON AS
+$$
+    SELECT json_object_agg(row.key, row.position) FROM (
+        SELECT key, pos as position FROM SCHEMA_CATALOG.exemplar_label_key_position
+            WHERE metric_name=metric_name_text GROUP BY metric_name, key, pos ORDER BY pos
+        ) AS row
+$$
+LANGUAGE SQL;
+GRANT EXECUTE ON FUNCTION SCHEMA_CATALOG.get_exemplar_label_key_positions(TEXT) TO prom_writer;
+GRANT EXECUTE ON FUNCTION SCHEMA_CATALOG.get_exemplar_label_key_positions(TEXT) TO prom_reader;
+
+-- creates exemplar table in prom_data_exemplar schema if the table does not exists. This function
+-- must be called after the metric is created in _prom_catalog.metric as it utilizes the table_name
+-- from the metric table. It returns true if the table was created.
+CREATE OR REPLACE FUNCTION SCHEMA_CATALOG.create_exemplar_table_if_not_exists(metric_name_text TEXT)
+RETURNS BOOLEAN
+AS
+$$
+DECLARE
+    table_name_fetched TEXT;
+    metric_name_fetched TEXT;
+BEGIN
+    SELECT metric_name, table_name INTO metric_name_fetched, table_name_fetched FROM SCHEMA_CATALOG.metric WHERE metric_name=metric_name_text;
+    IF table_name_fetched IS NULL THEN
+        -- metric table entry does not exists in SCHEMA_CATALOG.metric, hence we cannot create. Error out.
+        -- Note: even though we can create an entry from here, we should not as it keeps the approach systematic.
+        RAISE EXCEPTION 'SCHEMA_CATALOG.metric does not contain the table entry for % metric', metric_name_text;
+    END IF;
+    -- check if table is already created.
+    IF (
+        SELECT count(table_name) > 0 FROM SCHEMA_CATALOG.exemplar WHERE metric_name=metric_name_text
+    ) THEN
+        RETURN FALSE;
+    END IF;
+    -- table does not exists. Let's create it.
+    EXECUTE FORMAT('CREATE TABLE SCHEMA_DATA_EXEMPLAR.%I (time TIMESTAMPTZ NOT NULL, series_id BIGINT NOT NULL, exemplar_label_values SCHEMA_PROM.label_value_array, value DOUBLE PRECISION NOT NULL) WITH (autovacuum_vacuum_threshold = 50000, autovacuum_analyze_threshold = 50000)',
+        table_name_fetched);
+    EXECUTE format('GRANT SELECT ON TABLE SCHEMA_DATA_EXEMPLAR.%I TO prom_reader', table_name_fetched);
+    EXECUTE format('GRANT SELECT, INSERT ON TABLE SCHEMA_DATA_EXEMPLAR.%I TO prom_writer', table_name_fetched);
+    EXECUTE format('GRANT SELECT, INSERT, UPDATE, DELETE ON TABLE SCHEMA_DATA_EXEMPLAR.%I TO prom_modifier', table_name_fetched);
+    EXECUTE format('CREATE UNIQUE INDEX exemplar_index%s ON SCHEMA_DATA_EXEMPLAR.%I (series_id, time) INCLUDE (value)',
```

Comment:
We cannot put schema name in a index.

https://www.postgresql.org/docs/9.1/sql-createindex.html

`No schema name can be included here; the index is always created in the same schema as its parent table.`

Reply:
Also, i am not sure why GIN should be added, since I don't think that is possible on bigint due to primitive type, unless using `btree_gin` extension.